src/tests/e2etest/CommandRunnerTests.cs

```diff
@@ -47,46 +52,208 @@ public partial class CommandStatus
+            int timeoutSeconds = 30;
+            bool timeOut = ((DateTime.Now - startTime).TotalSeconds > timeoutSeconds);
+            while (!timeOut)
             {
-                var deserializedTwin = JsonSerializer.Deserialize<CommandStatus>(GetTwin().Properties.Reported["CommandRunner"]["CommandStatus"].ToString());
-                // Wait until our random commandId is present in reported properties
-                DateTime startTime = DateTime.Now;
-                while(deserializedTwin.CommandId != random && (DateTime.Now - startTime).TotalSeconds < 30)
+                if (IsCommandStatusUpdated(expectedCommandStatus, ignoreTextResult) && IsResponseCodeUpdated(expectedResponseCode))
                 {
-                    Console.WriteLine("[CommandRunnerTests] waiting for commandId to match...");
-                    Task.Delay(twinRefreshIntervalMs).Wait();
-                    deserializedTwin = JsonSerializer.Deserialize<CommandStatus>(GetNewTwin().Properties.Reported["CommandRunner"]["CommandStatus"].ToString());
+                    break;
                 }
 
-                AreEqualByJson(expectedCommandStatus, deserializedTwin);
-            }
-            else
-            {
-                Assert.Fail("Timeout for updating twin - RunCommand");
+                Task.Delay(twinRefreshIntervalMs).Wait();
+                timeOut = ((DateTime.Now - startTime).TotalSeconds > timeoutSeconds);
             }
+
+            return IsCommandStatusUpdated(expectedCommandStatus, ignoreTextResult) && IsResponseCodeUpdated(expectedResponseCode);
+        }
+
+        public bool IsCommandStatusUpdated(Object expectedCommandStatus, bool ignoreTextResult)
+        {
+            var deserializedTwin = JsonSerializer.Deserialize<CommandStatus>(GetNewTwin().Properties.Reported[ComponentName][PropertyCommandStatus].ToString());
+            deserializedTwin.TextResult = ignoreTextResult ? "" : deserializedTwin.TextResult;
+            return AreJsonObjectsEqual(expectedCommandStatus, deserializedTwin);
+        }
+
+        public bool IsResponseCodeUpdated(int expectedResponseCode)
+        {
+            var responseObject = JsonSerializer.Deserialize<ResponseCode>(GetTwin().Properties.Reported[ComponentName][PerpertyCommandArguments].ToString());
+            return responseObject.ac == expectedResponseCode;
+        }
+
+        [Test]
+        public void CommandRunnerTest_RunCommand_Echo_HelloWorld()
+        {
+            AssertModuleConnected();
+            // Run an echo Hello World command
+            (CommandArguments desiredCommand, CommandStatus expectedCommandStatus) = GenerateEchoHelloWorldCommand();
+            UpdateDesiredProperties(ComponentName, desiredCommand);
+            // Verify the command status (should be succeeded)
+            AreReportedPropertiesUpdated(expectedCommandStatus, responseStatusSuccess, false);
+        }
+
+        [Test]
+        public void CommandRunnerTest_Same_CommandId()
```

Comment:
Is it possible to also test this after OSConfig is restarted? This would test that the last commandIt is cached correctly to disk.

Reply:
Last command result is persisted as CommandId and numeric result. It can be used for this purpose.

Can the test reboot the device and then try to reboot again using the same CommandId?

This would be a test for the infinite loop of death conditions

Reply:
On second thought, this currently cannot be done since CommandRunner is only caching one command to disk specifically for reboot. We can add a test like I described (same commandId after reboot) later on in the future if/when CommandRunner gets updated.

Reply:
All last commands are persisted. Not just reboot. This test can be done if the test can reboot the device under test and continue all within same test session.

Can we do this?